File: plugin/src/main/java/org/jukeboxmc/plugin/perms/manager/GroupManagerImpl.java
package org.jukeboxmc.plugin.perms.manager;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.Nullable;
import org.jukeboxmc.api.config.Config;
import org.jukeboxmc.plugin.perms.api.manager.GroupManager;
import org.jukeboxmc.plugin.perms.api.model.Group;
import org.jukeboxmc.plugin.perms.repository.GroupsRepository;

/**
 * @author Kaooot
 * @version 1.0
 */
@Singleton
@RequiredArgsConstructor(onConstructor_ = @Inject)
public class GroupManagerImpl implements GroupManager {

    private final GroupsRepository repository;
    private final Config config;

    @Override
    public boolean createGroup(Group group) {
        return this.repository.insertGroup(group);
    }

    @Override
    public Optional<Group> groupById(String identifier) {
        return Optional.ofNullable(this.repository.selectGroup(identifier));
    }

    @Override
    public Set<Group> groups() {
        return this.repository.selectGroups();
    }

    @Override
    public void deleteGroup(String identifier) {
        this.repository.deleteGroup(identifier);
    }

    @Override
    public CompletableFuture<Boolean> addGroup(UUID uuid, String groupId) {
        return this.group(uuid).thenApply(group -> {
            if (group == null || group.id().equalsIgnoreCase(groupId)) {
                return false;
            }

            this.repository.addUserGroup(uuid, groupId);

            return true;
        });
    }

    @Override
    public CompletableFuture<Group> group(UUID uuid) {
        return this.repository.selectUserGroup(uuid);
    }

    @Override
    public CompletableFuture<Set<Group>> activeGroups(UUID uuid) {
        return this.repository.activeUserGroups(uuid);
    }

    @Override
    public @Nullable Group cachedGroup(UUID uuid) {
        return this.groupById(this.repository.userGroups().get(uuid)).orElse(null);
    }

    @Override
    public CompletableFuture<Boolean> removeGroup(UUID uuid, String groupId) {
        return this.activeGroups(uuid).thenApply(groups -> {
            if (groupId.equalsIgnoreCase(this.config.getString("fallback_group"))) {
                return false;
            }

            for (final Group group : groups) {
                if (!group.id().equalsIgnoreCase(groupId)) {
                    continue;
                }

                this.repository.removeUserGroup(uuid, group.id());

                return true;
            }

            return false;
        });
    }
}

File: plugin/src/main/kotlin/org/jukeboxmc/plugin/perms/manager/GroupManagerImpl.kt
package org.jukeboxmc.plugin.perms.manager

import org.jukeboxmc.api.config.Config
import org.jukeboxmc.plugin.perms.api.manager.GroupManager
import org.jukeboxmc.plugin.perms.api.model.Group
import org.jukeboxmc.plugin.perms.repository.GroupsRepository
import java.util.*
import java.util.concurrent.CompletableFuture

/**
 * @author Kaooot
 * @version 1.0
 */
class GroupManagerImpl(private val repository: GroupsRepository, private val config: Config) : GroupManager {

    override fun createGroup(group: Group): Boolean {
        return this.repository.insertGroup(group)
    }

    override fun groupById(identifier: String): Optional<Group> {
        return Optional.ofNullable(this.repository.selectGroup(identifier))
    }

    override fun groups(): Set<Group> {
        return this.repository.selectGroups()
    }

    override fun deleteGroup(identifier: String) {
        this.repository.deleteGroup(identifier)
    }

    override fun addGroup(uuid: UUID, group: String): CompletableFuture<Boolean> {
        return this.group(uuid).thenApply { g: Group? ->
            if (g == null || g.id().equals(group, ignoreCase = true)) {
                return@thenApply false
            }

            this.repository.addUserGroup(uuid, group)

            true
        }
    }

    override fun group(uuid: UUID): CompletableFuture<Group?> {
        return this.repository.selectUserGroup(uuid)
    }

    override fun activeGroups(uuid: UUID): CompletableFuture<Set<Group>> {
        return this.repository.activeUserGroups(uuid)
    }

    override fun cachedGroup(uuid: UUID): Group {
        return this.groupById(this.repository.userGroups()[uuid]!!).orElse(null)
    }

    override fun removeGroup(uuid: UUID, group: String): CompletableFuture<Boolean> {
        return this.activeGroups(uuid).thenApply { groups: Set<Group> ->
            if (group.equals(this.config.getString("fallback_group"), ignoreCase = true)) {
                return@thenApply false
            }

            for (g in groups) {
                if (!g.id().equals(group, ignoreCase = true)) {
                    continue
                }

                this.repository.removeUserGroup(uuid, g.id())

                return@thenApply true
            }

            false
        }
    }
}